src/services/bundle/bundle.ts

```diff
@@ -5,13 +5,28 @@ class Bundle {
   async publish(path: string, channel: string) {
     core.exportVariable('CHARMCRAFT_AUTH', core.getInput('credentials'));
     process.chdir(path);
-    await exec.exec('juju-bundle', [
-      'publish',
-      '--destructive-mode',
-      '--serial',
-      '--release',
-      channel,
-    ]);
+
+    const result = await exec.getExecOutput('charmcraft', ['pack']);
+    if (result.exitCode === 0) {
+      const lastLine = result.stderr.trim().split('\n').pop();
+      if (lastLine) {
+        const bundleName = lastLine.split(' ')[1];
+        await exec.exec('charmcraft', [
+          'upload',
+          bundleName,
+          '--release',
+          channel,
+        ]);
+      } else {
+        throw new Error(
+          'Failed to extract the bundle name from the output of charmcraft pack.',
```

Comment:
If `lastLine` is empty, I think it probably means that there is no stderr captured when running charmcraft pack, right? Maybe mentioning that it "Failed to extract the bundle name from output" is a bit misleading.

On a second thought, this is not entirely false, since there is no output, thus we indeed failed to extract it. Do you think it would make sense to clarify this error message a bit more?

Reply:
>This means that this error should never be reached, but the if else block is required because otherwise compiling the TypeScript produces an error. 

Yes +1. What you could do is if(!lastLine){ throwError..} and then the rest of the code as you did with the other `if..else` statement. But that's a nit. Feel free to keep the `if..else` if you like that better. 

>But then again, I am not 100% sure that output is always produced on stderr.

I played around a bit and indeed, I couldn't find a case where sth was produced in stdout. We can always ask a question in charmcraft channel just to verify this. WDYT?

>Is there a more appropriate structure for this situation in TypeScript? I'm thinking something like the "?" operator or the "unwrap()" function that exists in Rust.

There's this [operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Conditional_operator) but I think that using an if here is more than fine. My conern was regarding the message but since "the command should always output at least 1 non-empty line to stderr", then this LGTM.

Reply:
Ok, turns out that as of the time of writing, `charmcraft pack` uses a `progress` function that always outputs the message to stderr. Specifically, it's in [these lines](https://github.com/canonical/craft-application/blob/276d4f375e0607f6c46dd283798cc4d11a3537f3/craft_application/commands/lifecycle.py#L378-L384). In this case, working with stderr is OK (at least until [this issue](https://github.com/canonical/craft-application/issues/434) is completed).